File: archive/canvas-api/src/main/java/com/instructure/canvasapi/api/RecipientAPI.java
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi.api;

import com.instructure.canvasapi.model.Recipient;
import com.instructure.canvasapi.utilities.APIHelpers;
import com.instructure.canvasapi.utilities.CanvasCallback;

import retrofit.Callback;
import retrofit.http.EncodedQuery;
import retrofit.http.GET;
import retrofit.http.Path;
import retrofit.http.Query;


public class RecipientAPI extends BuildInterfaceAPI {

    interface RecipientsInterface {
        @GET("/search/recipients?synthetic_contexts=1")
        void getFirstPageRecipientsList(@Query("search") String searchTerm, @EncodedQuery("context")String context, Callback<Recipient[]> callback);

        @GET("/search/recipients?synthetic_contexts=1")
        void getFirstPageRecipientsListNoContext(@Query("search") String searchTerm, Callback<Recipient[]> callback);

        @GET("/{next}")
        void getNextPageRecipientsList(@Path(value = "next", encode = false) String nextURL, Callback<Recipient[]> callback);
    }

    /////////////////////////////////////////////////////////////////////////
    // API Calls
    /////////////////////////////////////////////////////////////////////////

    public static void getFirstPageRecipients(String search, String context, CanvasCallback<Recipient[]> callback) {
        if (APIHelpers.paramIsNull(callback, search, context)) { return; }

        if(context.trim().equals("")){
            buildCacheInterface(RecipientsInterface.class, callback).getFirstPageRecipientsListNoContext(search,callback);
            buildInterface(RecipientsInterface.class, callback).getFirstPageRecipientsListNoContext(search,callback);
        } else {
            buildCacheInterface(RecipientsInterface.class, callback).getFirstPageRecipientsList(search,context,callback);
            buildInterface(RecipientsInterface.class, callback).getFirstPageRecipientsList(search,context,callback);
        }
    }

    public static void getNextPageRecipients(String nextURL, CanvasCallback<Recipient[]> callback){
        if (APIHelpers.paramIsNull(callback, nextURL)) { return; }

        callback.setIsNextPage(true);
        buildCacheInterface(RecipientsInterface.class, callback, false).getNextPageRecipientsList(nextURL,callback);
        buildInterface(RecipientsInterface.class, callback, false).getNextPageRecipientsList(nextURL,callback);
    }
}

File: libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/apis/RecipientAPI.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi2.apis

import com.instructure.canvasapi2.StatusCallback
import com.instructure.canvasapi2.builders.RestBuilder
import com.instructure.canvasapi2.builders.RestParams
import com.instructure.canvasapi2.models.Recipient

import retrofit2.Call
import retrofit2.http.GET
import retrofit2.http.Query
import retrofit2.http.Url


object RecipientAPI {

    internal interface RecipientInterface {
        @GET("search/recipients?synthetic_contexts=1")
        fun getFirstPageRecipientList(@Query("search") searchQuery: String?, @Query(value = "context", encoded = true) context: String): Call<List<Recipient>>

        @GET("search/recipients")
        fun getFirstPageRecipientListNoSyntheticContexts(@Query("search") searchQuery: String?, @Query(value = "context", encoded = true) context: String): Call<List<Recipient>>

        @GET
        fun getNextPageRecipientList(@Url url: String): Call<List<Recipient>>
    }

    fun getRecipients(searchQuery: String?, context: String, callback: StatusCallback<List<Recipient>>, adapter: RestBuilder, params: RestParams) {
        if (StatusCallback.isFirstPage(callback.linkHeaders)) {
            callback.addCall(adapter.build(RecipientInterface::class.java, params).getFirstPageRecipientList(searchQuery, context)).enqueue(callback)
        } else if (callback.linkHeaders != null && StatusCallback.moreCallsExist(callback.linkHeaders)) {
            callback.addCall(adapter.build(RecipientInterface::class.java, params).getNextPageRecipientList(callback.linkHeaders!!.nextUrl!!)).enqueue(callback)
        }
    }

    fun getFirstPageRecipients(forceNetwork: Boolean, searchQuery: String?, context: String, adapter: RestBuilder, callback: StatusCallback<List<Recipient>>) {
        val params = RestParams(usePerPageQueryParam = true, isForceReadFromNetwork = forceNetwork)
        callback.addCall(adapter.build(RecipientInterface::class.java, params).getFirstPageRecipientList(searchQuery, context)).enqueue(callback)
    }

    // Synthetic contexts == groups and sections, so this search will return only users
    fun getFirstPageRecipientsNoSyntheticContexts(forceNetwork: Boolean, searchQuery: String?, context: String, adapter: RestBuilder, callback: StatusCallback<List<Recipient>>) {
        val params = RestParams(usePerPageQueryParam = true, isForceReadFromNetwork = forceNetwork)
        callback.addCall(adapter.build(RecipientInterface::class.java, params).getFirstPageRecipientListNoSyntheticContexts(searchQuery, context)).enqueue(callback)
    }

    fun getNextPageRecipients(forceNetwork: Boolean, nextUrl: String, adapter: RestBuilder, callback: StatusCallback<List<Recipient>>) {
        val params = RestParams(usePerPageQueryParam = true, isForceReadFromNetwork = forceNetwork)
        adapter.build(RecipientInterface::class.java, params).getNextPageRecipientList(nextUrl).enqueue(callback)
    }
}
